src/charm.py

```diff
@@ -224,57 +207,62 @@ def _bessd_config_file_content_matches(self, content: str) -> bool:
-        """Handles events for configuring workloads.
-
-        Args:
-            event: EventBase
-        """
+    def _on_config_changed(self, event: EventBase):
+        """Handler for config changed events."""
+        if not self.unit.is_leader():
+            return
         if invalid_configs := self._get_invalid_configs():
             self.unit.status = BlockedStatus(
                 f"The following configurations are not valid: {invalid_configs}"
             )
             return
-        if not self._kubernetes_multus.is_ready():
-            self.unit.status = WaitingStatus("Waiting for statefulset to be patched")
-            return
         if not self._bessd_container.can_connect():
-            self.unit.status = WaitingStatus(
-                "Waiting to be able to connect to the `bessd` container"
-            )
+            self.unit.status = WaitingStatus("Waiting for bessd container to be ready")
+            event.defer()
             return
-        if not self._routectl_container.can_connect():
-            self.unit.status = WaitingStatus(
-                "Waiting to be able to connect to the `routectl` container"
-            )
+        if not self._pfcp_agent_container.can_connect():
+            self.unit.status = WaitingStatus("Waiting for pfcp agent container to be ready")
+            event.defer()
             return
-        if not self._web_container.can_connect():
-            self.unit.status = WaitingStatus(
-                "Waiting to be able to connect to the `web` container"
-            )
+        self._on_bessd_pebble_ready(event)
+        self._on_pfcp_agent_pebble_ready(event)
+
+    def _on_bessd_pebble_ready(self, event: EventBase) -> None:
+        """Handle Pebble ready event."""
+        if not self.unit.is_leader():
             return
-        if not self._pfcp_agent_container.can_connect():
-            self.unit.status = WaitingStatus(
-                "Waiting to be able to connect to the `pfcp-agent` container"
-            )
+        if not self._kubernetes_multus.is_ready():
+            self.unit.status = WaitingStatus("Waiting for Multus to be ready")
+            event.defer()
             return
         if not self._bessd_container.exists(path=BESSD_CONTAINER_CONFIG_PATH):
             self.unit.status = WaitingStatus("Waiting for storage to be attached")
             event.defer()
             return
-        self._update_fiveg_n3_relation_data(event)
         self._configure_bessd_workload()
-        self._configure_routectl_workload()
-        self._configure_web_workload()
-        self._configure_pfcp_agent_workoad()
+        self._set_unit_status()
+
+    def _on_pfcp_agent_pebble_ready(self, event: EventBase) -> None:
+        """Handle pfcp agent Pebble ready event."""
+        if not self.unit.is_leader():
+            return
+        if not service_is_running_on_container(self._bessd_container, self._bessd_service_name):
+            self.unit.status = WaitingStatus(
```

Comment:
Why do we need to wait for `bessd`?

Reply:
It's a pre-requisite for the service to start correctly